time: use clock_gettime on macos

tamird opened the pull request:
Replace `gettimeofday` with `clock_gettime(CLOCK_REALTIME)` on:

```
all(target_os = "macos", not(target_arch = "aarch64")),
    target_os = "ios",
    target_os = "watchos",
    target_os = "tvos"
))]
```

`gettimeofday` was first used in
https://github.com/time-rs/time/commit/cc367edd953e72756ed6f0980918795c11e469b1
which predated the introduction of `clock_gettime` support in macOS
10.12 Sierra which became the minimum supported version in
58bbca958d917a89124da248735926f86c59a149.

Replace `mach_{absolute_time,timebase_info}` with
`clock_gettime(CLOCK_REALTIME)` on:

```
all(target_os = "macos", not(target_arch = "aarch64")),
    target_os = "ios",
    target_os = "watchos",
    target_os = "tvos"
))]
```

`mach_{absolute_time,timebase_info}` were first used in
https://github.com/time-rs/time/commit/cc367edd953e72756ed6f0980918795c11e469b1
which predated the introduction of `clock_gettime` support in macOS
10.12 Sierra which became the minimum supported version in
58bbca958d917a89124da248735926f86c59a149.

Note that this change was made for aarch64 in
5008a317ce8e508c390ed12bff281f307313376e which predated 10.12 becoming
the minimum supported version. The discussion took place in
https://github.com/rust-lang/rust/issues/91417 and in particular
https://github.com/rust-lang/rust/issues/91417#issuecomment-992151582
and
https://github.com/rust-lang/rust/issues/91417#issuecomment-1033048064
are relevant.

tamird committed:
time: cfg(any(x)) is the same as cfg(x)

This was left over in c043a0e7d6f347f9b2c7f08c5b3a179470e0f0c5.

tamird committed:
time: use clock_gettime on macos

Replace `gettimeofday` with `clock_gettime(CLOCK_REALTIME)` on:

```
all(target_os = "macos", not(target_arch = "aarch64")),
    target_os = "ios",
    target_os = "watchos",
    target_os = "tvos"
))]
```

`gettimeofday` was first used in
https://github.com/time-rs/time/commit/cc367edd953e72756ed6f0980918795c11e469b1
which predated the introduction of `clock_gettime` support in macOS
10.12 Sierra which became the minimum supported version in
58bbca958d917a89124da248735926f86c59a149.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6342068134/job/17227062179) [(plain)](https://github.com/rust-lang/rust/commit/9033c8331505477e57dc4c302e1a466877486ab8/checks/17227062179/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=tamird
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_17a0fe95-6fe4-465d-a197-886d23eae07d
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=b7f56bc195d34f40439ddfcdfa8c64a352628ef5
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_17a0fe95-6fe4-465d-a197-886d23eae07d
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_17a0fe95-6fe4-465d-a197-886d23eae07d
GITHUB_TRIGGERING_ACTOR=tamird
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116238/merge
GITHUB_WORKFLOW_SHA=b7f56bc195d34f40439ddfcdfa8c64a352628ef5
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
   Compiling miniz_oxide v0.7.1
   Compiling std_detect v0.1.5 (/checkout/library/stdarch/crates/std_detect)
   Compiling hashbrown v0.14.0
   Compiling addr2line v0.21.0
error[E0412]: cannot find type `__timespec64` in module `super`
   --> library/std/src/sys/unix/time.rs:102:69
    |
102 |             weak!(fn __clock_gettime64(libc::clockid_t, *mut super::__timespec64) -> libc::c_int);
    |                                                                     ^^^^^^^^^^^^ not found in `super`
help: consider importing this struct
    |
    |
1   + use crate::sys::time::__timespec64;
    |
help: if you import `__timespec64`, refer to it directly
    |
102 -             weak!(fn __clock_gettime64(libc::clockid_t, *mut super::__timespec64) -> libc::c_int);
102 +             weak!(fn __clock_gettime64(libc::clockid_t, *mut __timespec64) -> libc::c_int);

For more information about this error, try `rustc --explain E0412`.
error: could not compile `std` (lib) due to previous error
Build completed unsuccessfully in 0:00:42
```

</details>

tamird commented:
r? @joshtriplett

tamird committed:
time: use clock_gettime on macos

Replace `mach_{absolute_time,timebase_info}` with
`clock_gettime(CLOCK_REALTIME)` on:

```
all(target_os = "macos", not(target_arch = "aarch64")),
    target_os = "ios",
    target_os = "watchos",
    target_os = "tvos"
))]
```

`mach_{absolute_time,timebase_info}` were first used in
https://github.com/time-rs/time/commit/cc367edd953e72756ed6f0980918795c11e469b1
which predated the introduction of `clock_gettime` support in macOS
10.12 Sierra which became the minimum supported version in
58bbca958d917a89124da248735926f86c59a149.

Note that this change was made for aarch64 in
5008a317ce8e508c390ed12bff281f307313376e which predated 10.12 becoming
the minimum supported version. The discussion took place in
https://github.com/rust-lang/rust/issues/91417 and in particular
https://github.com/rust-lang/rust/issues/91417#issuecomment-992151582
and
https://github.com/rust-lang/rust/issues/91417#issuecomment-1033048064
are relevant.

tamird commented:
cc @thomcc and @hkratz who were involved in https://github.com/rust-lang/rust/issues/91417 and may generally be knowledgeable about this arcana.

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6344569950/job/17234965714) [(plain)](https://github.com/rust-lang/rust/commit/1945d234b9a10cf22f854d1277a5d9118ca19408/checks/17234965714/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=tamird
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_262cae98-feb1-4600-b9ef-157624ab4f22
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=c24fa37514fcb471b794ff435ab4ccbaf59c947f
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_262cae98-feb1-4600-b9ef-157624ab4f22
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_262cae98-feb1-4600-b9ef-157624ab4f22
GITHUB_TRIGGERING_ACTOR=tamird
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116238/merge
GITHUB_WORKFLOW_SHA=c24fa37514fcb471b794ff435ab4ccbaf59c947f
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---

Sending build context to Docker daemon  432.1kB

Step 1/15 : FROM ubuntu:22.04
Head "https://registry-1.docker.io/v2/library/ubuntu/manifests/22.04": Get "https://auth.docker.io/token?account=githubactions&scope=repository%3Alibrary%2Fubuntu%3Apull&service=registry.docker.io": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
            BuildKit is currently disabled; enable it by removing the DOCKER_BUILDKIT=0
            environment-variable.


Sending build context to Docker daemon  432.1kB

Step 1/15 : FROM ubuntu:22.04
Head "https://registry-1.docker.io/v2/library/ubuntu/manifests/22.04": Get "https://auth.docker.io/token?account=githubactions&scope=repository%3Alibrary%2Fubuntu%3Apull&service=registry.docker.io": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
            BuildKit is currently disabled; enable it by removing the DOCKER_BUILDKIT=0
            environment-variable.

---

Sending build context to Docker daemon  432.1kB

Step 1/15 : FROM ubuntu:22.04
Head "https://registry-1.docker.io/v2/library/ubuntu/manifests/22.04": Get "https://auth.docker.io/token?account=githubactions&scope=repository%3Alibrary%2Fubuntu%3Apull&service=registry.docker.io": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

thomcc commented on library/std/src/sys/unix/time.rs:
```diff
@@ -216,209 +248,34 @@ impl From<__timespec64> for Timespec {
-        t: Timespec,
+impl Instant {
+    pub fn now() -> Instant {
+        Instant { t: Timespec::now(libc::CLOCK_MONOTONIC) }
```
This is a semantic change from the previous version (which used mach_absolute_time, which would be equivalent to CLOCK_UPTIME_RAW).

I would prefer we not make such a semantic change without consideration.

tamird replied:
Yep, and we can keep the previous behavior. I called this out in the commit message, but I'll say it again here: I could not figure out from #91417 why the move to `CLOCK_UPTIME_RAW` was made.

thomcc replied:
It's because it's equivalent to `mach_absolute_time`.

thomcc replied:
To be more specific, there's a real way that `mach_absolute_time` is the true monotonic kernel timestamp. I used to think we should switch away from it, but so much of the kernel's internals (and even public APIs) are based around it that I think we should have a good reason to do so.

tamird replied:
Sounds good. I've updated this, have another look.

thomcc commented:
I probably have more experience here than josh, so I'll take the review unless they object.

r? thomcc

joshtriplett commented:
@thomcc By all means. I'm familiar with clock_gettime but I'm not a macOS expert at all.

thomcc commented:
Yeah, I'm very familiar with the details of the macOS implementation of it.

tamird commented:
@thomcc any further comments?

thomcc commented:
Sorry, I have to verify which clock the new apple guidance wants us to use. I have time this weekend to do it (I had hoped to do it this past weekend, but real life got in the way), but perhaps not before then.

thomcc commented:
Indeed, it does not.

tamird commented:
Friendly ping @thomcc - have you had a chance to think about this?

tamird commented:
Hi @thomcc, are you still interested in reviewing this PR?

thomcc commented:
Yeah this is fine for now.

@bors r+

bors commented:
:pushpin: Commit bc300102d4388860afa895149c1f52393a7aa964 has been approved by `thomcc`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=thomcc bc300102d4388860afa895149c1f52393a7aa964 -->
<!-- homu: {"type":"Approved","sha":"bc300102d4388860afa895149c1f52393a7aa964","approver":"thomcc","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit bc300102d4388860afa895149c1f52393a7aa964 with merge e918db897df677d9bbda64d82ebbb8f1bc9e557f...
<!-- homu: {"type":"BuildStarted","head_sha":"bc300102d4388860afa895149c1f52393a7aa964","merge_sha":"e918db897df677d9bbda64d82ebbb8f1bc9e557f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6622268755/job/17989413768)
Approved by: thomcc
Pushing e918db897df677d9bbda64d82ebbb8f1bc9e557f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"thomcc","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6622268755/job/17989413768"},"merge_sha":"e918db897df677d9bbda64d82ebbb8f1bc9e557f"} -->

rust-timer commented:
Finished benchmarking commit (e918db897df677d9bbda64d82ebbb8f1bc9e557f): [comparison URL](https://perf.rust-lang.org/compare.html?start=f1a5ce19f5aa0cf61ed7b9f75b30e610befeed72&end=e918db897df677d9bbda64d82ebbb8f1bc9e557f&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=f1a5ce19f5aa0cf61ed7b9f75b30e610befeed72&end=e918db897df677d9bbda64d82ebbb8f1bc9e557f&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.6% | [-3.6%, -3.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -3.6% | [-3.6%, -3.6%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f1a5ce19f5aa0cf61ed7b9f75b30e610befeed72&end=e918db897df677d9bbda64d82ebbb8f1bc9e557f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 8.2% | [8.2%, 8.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f1a5ce19f5aa0cf61ed7b9f75b30e610befeed72&end=e918db897df677d9bbda64d82ebbb8f1bc9e557f&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 6.3%  | [1.5%, 11.6%]  | 6     |
| Improvements ✅ <br /> (primary)   | -3.4% | [-3.4%, -3.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -3.4% | [-3.4%, -3.4%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=f1a5ce19f5aa0cf61ed7b9f75b30e610befeed72&end=e918db897df677d9bbda64d82ebbb8f1bc9e557f&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: missing data
**Artifact size**: 304.36 MiB -> 304.36 MiB (-0.00%)
